File: apps/api/internal/api/aggregations/dto.go
package aggregations

import (
	"time"
)

type UserResponseDto struct {
	ID           string    `json:"id"`
	Username     string    `json:"username"`
	FullName     string    `json:"fullName"`
	Gender       *string   `json:"gender"`
	ProfileImage *string   `json:"profileImage"`
	CreatedAt    time.Time `json:"createdAt"`
}

type HServiceResponseDto struct {
	ID               string          `json:"id"`
	UserID           string          `json:"userId"`
	User             UserResponseDto `json:"user"`
	Title            string          `json:"title"`
	Slug             string          `json:"slug"`
	Description      string          `json:"description"`
	Category         int32           `json:"category"`
	Price            float64         `json:"price"`
	PriceUnit        string          `json:"priceUnit"`
	PriceTimespan    string          `json:"priceTimespan"`
	IsOnline         bool            `json:"isOnline"`
	Url              *string         `json:"url"`
	Location         string          `json:"location"`
	DeliveryTime     int32           `json:"deliveryTime"`
	DeliveryTimespan string          `json:"deliveryTimespan"`
	TotalPoints      int64           `json:"totalPoints"`
	TotalVotes       int32           `json:"totalVotes"`
	Media            map[string]any  `json:"media"`
	CreatedAt        time.Time       `json:"createdAt"`
	UpdatedAt        time.Time       `json:"updatedAt"`
}

type GetHomeAggregationsResponseDto struct {
	New       []HServiceResponseDto `json:"new"`
	Popular   []HServiceResponseDto `json:"popular"`
	Featured  []HServiceResponseDto `json:"featured"`
	Favorites []HServiceResponseDto `json:"favorites"`
}

File: web/src/lib/dto.ts
export type GetMeResponseDto = {
  id: string;
  email: string;
  username: string;
  fullName: string;
  googleId: string | null;
  isEmailVerified: boolean;
  isActive: boolean;
  role: string;
  gender: string | null;
  profileImage: string | null;
  lastLogin: string;
  createdAt: string;
  updatedAt: string;
};

export type UploadImageType = 'hservices' | 'profile' | 'reviews';

export type CreateHServiceRequestDto = {
  title: string;
  description: string;
  category: number;
  price: number;
  priceUnit: string;
  priceTimespan: string;
  isOnline: boolean;
  url: string;
  location: string;
  deliveryTime: number;
  deliveryTimespan: string;
  media: string;
};

export type UserDto = GetUserProfileByUsernameResponseDto;

export type HServiceResponseDto = {
  id: string;
  user: UserDto;
  userId: string;
  title: string;
  slug: string;
  description: string;
  category: number;
  price: number;
  priceUnit: string;
  priceTimespan: string;
  isOnline: boolean;
  url: string | null;
  location: string;
  deliveryTime: number;
  deliveryTimespan: string;
  totalPoints: number;
  totalVotes: number;
  media: MediaResponseDto;
  createdAt: string;
  updatedAt: string;
};

export type HServiceWithoutUserResponseDto = {
  id: string;
  userId: string;
  title: string;
  slug: string;
  description: string;
  category: number;
  price: number;
  priceUnit: string;
  priceTimespan: string;
  isOnline: boolean;
  url: string | null;
  location: string;
  deliveryTime: number;
  deliveryTimespan: string;
  totalPoints: number;
  totalVotes: number;
  media: MediaResponseDto;
  createdAt: string;
  updatedAt: string;
};

export type MediaResponseDto = {
  data: Media[];
};

export type Media = {
  type: 'image' | 'video';
  url: string;
  thumbnail: string;
  alt: string;
  caption: string | null;
  width: number;
  height: number;
};

export type GetUserProfileByUsernameResponseDto = {
  id: string;
  username: string;
  fullName: string;
  gender: string | null;
  profileImage: string | null;
  createdAt: string | null;
};

export type GetNewUploadUrlResponseDto = {
  key: string;
  url: string;
};

export type GetHomeAggregationsResponseDto = {
  new: HServiceResponseDto[];
  popular: HServiceResponseDto[];
  featured: HServiceResponseDto[];
  favorites: HServiceResponseDto[];
};

export type HServiceMetadataDto = {
  isFavorite: boolean;
  isBookmarked: boolean;
};

export type SendForgotPasswordEmailRequestDto = {
  email: string;
};

export type ResetPasswordRequestDto = {
  email: string;
  code: string;
  newPassword: string;
};

export type BookmarksResponseItemDto = {
  id: string;
  userId: string;
  hserviceId: string;
  createdAt: string;
  hservice: HServiceWithoutUserResponseDto;
};

export type Pagination = {
  page: number;
  pageSize: number;
  totalRecords: number;
  totalPages: number;
  hasPrevious: boolean;
  hasNext: boolean;
};

export type GetMyListsResponseDtoItem = {
  id: string;
  title: string;
  userId: string;
  createdAt: string;
  updatedAt: string;
};

export type GetMyListsResponseDto = GetMyListsResponseDtoItem[];

export type CreateListRequestDto = {
  title: string;
};

export type GetListByIdResponseDto = {
  id: string;
  title: string;
  userId: string;
  user: UserDto;
  items: ListItemDto[];
  createdAt: string;
  updatedAt: string;
};

export type ListItemDto = {
  id: string;
  listId: string;
  hserviceId: string;
  hservice: HServiceResponseDto;
  itemOrder: number;
};
